crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,618 @@
+        //   structures used for internal tracking: the only Python
+        //   references they contain are to the bits in these lists!
+        visit.call(&self.qubits)?;
+        visit.call(&self.clbits)?;
+        Ok(())
+    }
+
+    fn __clear__(&mut self) {
+        // Clear anything that could have a reference cycle.
+        self.data.clear();
+        self.qubits_native.clear();
+        self.clbits_native.clear();
+        self.qubit_indices_native.clear();
+        self.clbit_indices_native.clear();
+    }
+}
+
+impl CircuitData {
+    /// Converts a Python slice to a `Vec` of indices into
+    /// the instruction listing, [CircuitData.data].
+    fn convert_py_slice(&self, slice: &PySlice) -> PyResult<Vec<isize>> {
+        let indices = slice.indices(self.data.len().try_into().unwrap())?;
+        if indices.step > 0 {
+            Ok((indices.start..indices.stop)
+                .step_by(indices.step as usize)
+                .collect())
+        } else {
+            let mut out = Vec::with_capacity(indices.slicelength as usize);
+            let mut x = indices.start;
+            while x > indices.stop {
+                out.push(x);
+                x += indices.step;
+            }
+            Ok(out)
+        }
+    }
+
+    /// Converts a Python index to an index into the instruction listing,
+    /// or one past its end.
+    /// If the resulting index would be < 0, clamps to 0.
+    /// If the resulting index would be > len(data), clamps to len(data).
+    fn convert_py_index_clamped(&self, index: isize) -> usize {
+        let index = if index < 0 {
+            index + self.data.len() as isize
+        } else {
+            index
+        };
+        std::cmp::min(std::cmp::max(0, index), self.data.len() as isize) as usize
+    }
+
+    /// Converts a Python index to an index into the instruction listing.
+    fn convert_py_index(&self, index: isize) -> PyResult<usize> {
+        let index = if index < 0 {
+            index + self.data.len() as isize
+        } else {
+            index
+        };
+
+        if index < 0 || index >= self.data.len() as isize {
+            return Err(PyIndexError::new_err(format!(
+                "Index {:?} is out of bounds.",
+                index,
+            )));
+        }
+        Ok(index as usize)
+    }
+
+    /// Returns an [InternedInstruction] containing the original operation
+    /// of `elem` and [InternContext] indices of its `qubits` and `clbits`
+    /// fields.
+    fn intern_instruction(
+        &mut self,
+        py: Python<'_>,
+        elem: PyRef<CircuitInstruction>,
+    ) -> PyResult<InternedInstruction> {
+        let mut interned_bits =
+            |indices: &HashMap<BitAsKey, u32>, bits: &PyTuple| -> PyResult<IndexType> {
+                let args = bits
+                    .into_iter()
+                    .map(|b| {
+                        let key = BitAsKey::new(b)?;
+                        indices.get(&key).cloned().ok_or_else(|| {
```

Comment:
Hmm, probably. Is that better?

Reply:
Functionally no, for all the numeric builtin types in rust `.clone()` just returns `*self`. It's more just about being a bit more explicit. For me it's more just that when I see `clone()` it triggers the thought process that we're doing an expensive copy and should look at how/if it can be avoided.

Reply:
Turned out there's also a `copied` method for `Option`, so I switched to that in b10d8b48445b061cf54212c9cb20de46378b2bdc